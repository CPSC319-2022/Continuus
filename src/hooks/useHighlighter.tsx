import { useCallback } from "react";

/**
 * Custom react hook to put <mark></mark> tags around substring based on regex match
 *
 * Sources:
 * - https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex
 * - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp
 *
 * @returns JSX with regex matches wrapped in <mark> tag
 *
 */
export const useHighlighter = () => {
  return useCallback(
    (str: string, searchInput: string) =>
      searchInput
        ? str
            .replaceAll(
              new RegExp(
                searchInput.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"),
                "gi"
              ),
              (substring) => `@@${substring}@@`
            )
            .split("@@")
            .map((str, i) =>
              i % 2 === 0 ? (
                str
              ) : (
                <mark key={i} className="bg-sky-200 dark:bg-amber-200">
                  {str}
                </mark>
              )
            )
        : str,
    []
  );
};
