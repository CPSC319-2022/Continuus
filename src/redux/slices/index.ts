import { combineReducers } from "@reduxjs/toolkit";
import { reducer as postsReducer } from "./posts";

export const reducer = combineReducers({
  posts: postsReducer,
});
