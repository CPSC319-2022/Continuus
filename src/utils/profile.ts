/**
 * Returns a path to the profile page for the given profile. If the argument is undefined, redirect to the home page.
 *
 * @param profile the user id of the account
 *
 * @return the profile path if the profile is a string, otherwise the home path
 */
export const userPathToProfile = (profile: string | undefined): string => {
    if (profile === undefined) {
        return '/';
    }

    return '/profile/' + encodeURIComponent(profile);
}
