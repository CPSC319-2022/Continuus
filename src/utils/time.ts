export const timeAgo = (inputDate: Date): string => {
  const now = new Date();
  const diffInSeconds = Math.max(Math.floor(
    (now.getTime() - inputDate.getTime()) / 1000
  ), 0);

  if (diffInSeconds < 60) {
    return `${diffInSeconds} ${diffInSeconds === 1 ? "second" : "seconds"} ago`;
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 30) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  const diffInMonths = Math.floor(diffInDays / 30);
  if (diffInMonths < 12) {
    return `${diffInMonths} ${diffInMonths === 1 ? 'month' : 'months'} ago`;
  }

  const diffInYears = Math.floor(diffInMonths / 12);
  return `${diffInYears} ${diffInYears === 1 ? 'year' : 'years'} ago`;
};
