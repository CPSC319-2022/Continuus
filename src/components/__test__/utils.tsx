import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
