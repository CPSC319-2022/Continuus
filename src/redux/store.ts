import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { reducer } from "./slices";
// Create the root reducer independently to obtain the RootState type
export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof reducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore["dispatch"];
