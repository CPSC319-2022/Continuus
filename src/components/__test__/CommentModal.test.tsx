import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup, render, waitFor, screen } from "@testing-library/react";
import * as reduxHooks from "~/redux/hooks";
import { renderWithProviders } from "./utils";

const reader = {
  createdAt: new Date(),
  email: "email",
  id: "reader-id",
  image: "image",
  name: "reader",
  role: "READER",
  updatedAt: new Date(),
};

export const mockComments = [
  {
    id: "comment-1",
    blogPostId: "post-1",
    content: "test comment",
    createdAt: new Date(),
    updatedAt: new Date(),
    user: {
      name: "test user 2",
      id: "user-2",
      image: "image",
    },
  },
  {
    id: "comment-2",
    blogPostId: "post-1",
    content: "test comment 2",
    createdAt: new Date(),
    updatedAt: new Date(),
    user: {
      name: "test user 3",
      id: "user-3",
      image: "image",
    },
  },
];

const setup = (user: any = null, post: any = null) => {
  vi.doMock("next/router", () => ({
    useRouter: () => ({
      events: {
        on: vi.fn(),
      },
    }),
  }));

  vi.mock("next-auth/react", () => ({
    useSession: () => ({
      data: {
        expires: "1",
        user: { id: "test-user-id", email: "a", name: "hi", image: "c" },
      },
    }),
  }));

  vi.doMock("~/utils/api", () => ({
    api: {
      user: {
        currentUser: {
          useQuery: () => ({
            data: user,
            isLoading: false,
          }),
        },
        updateOne: {
          useMutation: vi.fn(),
        },
      },
      blogPost: {
        getOne: {
          useQuery: () => ({
            data: {
              id: "post-1",
              userId: "user-1",
              title: "title",
              updatedAt: new Date(),
              createdAt: new Date(),
              content: "content",
              comments: mockComments,
              user: {
                name: "test user",
                image: "imageUrl",
              },
            },
            isLoading: false,
          }),
        },
      },
      comment: {
        getUnique: {
          useQuery: () => ({
            data: {
              id: "comment-1",
              blogPostId: "post-1",
              content: "test comment",
              createdAt: new Date(),
              updatedAt: new Date(),
              user: {
                name: "test user 2",
                id: "user-2",
                image: "image",
              },
            },
            isLoading: false,
          }),
        },
        create: {
          useMutation: () => ({
            onSuccess: vi.fn(),
          }),
        },
        update: {
          useMutation: () => ({
            onSuccess: vi.fn(),
          }),
        },
        delete: {
          useMutation: () => ({
            onSuccess: vi.fn(),
          }),
        },
      },
      useContext: () => ({
        blogPost: {
          invalidate: vi.fn(),
        },
        comment: {
          count: {
            invalidate: vi.fn(),
          },
        },
      }),
    },
  }));
};

describe("Comment Modal: Snapshots", () => {
  const useSelectorMock = vi.spyOn(reduxHooks, "useAppSelector");
  const useDispatchMock = vi.spyOn(reduxHooks, "useAppDispatch");

  beforeEach(() => {
    cleanup();
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    vi.clearAllMocks();
  });

  it("render comment modal not logged in no selected post", async () => {
    setup();
    const { CommentModal } = await import("~/components/CommentModal");
    const body = renderWithProviders(<CommentModal />);
    expect(body.baseElement).toMatchSnapshot();
  });

  it("render comment modal not logged in with selected post", async () => {
    setup(null);
    useSelectorMock.mockReturnValue({
      posts: {
        selectedPost: "post-1",
      },
    });
    const { CommentModal } = await import("~/components/CommentModal");
    const body = renderWithProviders(<CommentModal />);
    expect(body.baseElement).toMatchSnapshot();
  });

  it("render comment modal logged in with selected post", async () => {
    setup(reader);
    useSelectorMock.mockReturnValue({
      posts: {
        selectedPost: "post-1",
      },
    });
    const { CommentModal } = await import("~/components/CommentModal");
    const body = renderWithProviders(<CommentModal />);
    expect(body.baseElement).toMatchSnapshot();
  });
});
