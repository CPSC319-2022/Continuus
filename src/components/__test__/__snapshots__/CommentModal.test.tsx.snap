// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Comment Modal: Snapshots > render comment modal logged in with selected post 1`] = `
<body
  aria-hidden="true"
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  />
  <div />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 max-h-[75vh] w-11/12 max-w-5xl -translate-x-1/2 -translate-y-1/2 overflow-y-auto overflow-x-hidden rounded-md border bg-white p-8 outline-none md:w-1/2"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="mb-4 flex w-full justify-between"
        >
          <div
            class="flex"
          >
            <div
              class="avatar self-center hover:scale-110 transition-all"
            >
               
              <a
                href="/profile/user-1"
              >
                <img
                  alt="avatar"
                  class="inline-block aspect-square rounded-[50%] "
                  src="imageUrl"
                  style="width: 2.5rem;"
                />
              </a>
            </div>
            <div
              class="ml-3"
            >
              <a
                href="/profile/user-1"
              >
                <p
                  class="text-lg font-bold hover:underline"
                >
                  test user
                </p>
              </a>
              <p
                class="text-sm text-gray-400"
              >
                0 seconds ago
              </p>
            </div>
          </div>
        </div>
        <p
          class="mb-3 text-2xl font-bold"
        >
          title
        </p>
        <div
          class="prose max-w-none mb-3"
        >
          <p>
            content
          </p>
        </div>
        <div
          class="flex w-full justify-end border-b mb-3"
        >
          <p
            class="float-right text-highlight-green no-underline hover:no-underline mb-3"
          >
            2
             Comments
          </p>
        </div>
        <div>
          <div
            class="relative m-3 rounded-md border-b p-4"
          >
            <div
              class="flex"
            >
              <div
                class="avatar self-center transition-all hover:scale-110"
              >
                 
                <a
                  href="/profile/user-2"
                >
                  <img
                    alt="avatar"
                    class="inline-block aspect-square rounded-[50%] "
                    src="image"
                    style="width: 2.5rem;"
                  />
                </a>
              </div>
              <div
                class="ml-3"
              >
                <a
                  href="/profile/user-2"
                >
                  <p
                    class="text-lg font-bold hover:underline"
                  >
                    test user 2
                  </p>
                </a>
                <p
                  class="text-gray-400 text-sm"
                >
                  0 seconds ago
                </p>
              </div>
            </div>
            <div
              class="break-words"
            >
              <p
                class="mt-4 w-full"
              >
                test comment
              </p>
            </div>
          </div>
          <div
            class="relative m-3 rounded-md border-b p-4"
          >
            <div
              class="flex"
            >
              <div
                class="avatar self-center transition-all hover:scale-110"
              >
                 
                <a
                  href="/profile/user-3"
                >
                  <img
                    alt="avatar"
                    class="inline-block aspect-square rounded-[50%] "
                    src="image"
                    style="width: 2.5rem;"
                  />
                </a>
              </div>
              <div
                class="ml-3"
              >
                <a
                  href="/profile/user-3"
                >
                  <p
                    class="text-lg font-bold hover:underline"
                  >
                    test user 3
                  </p>
                </a>
                <p
                  class="text-gray-400 text-sm"
                >
                  0 seconds ago
                </p>
              </div>
            </div>
            <div
              class="break-words"
            >
              <p
                class="mt-4 w-full"
              >
                test comment
              </p>
            </div>
          </div>
        </div>
        <div
          class="tooltip tooltip-info w-full"
          data-tip="Sign in to comment"
        >
          <input
            class="input-disabled cursor-not-allowedinput-bordered mb-3 w-full rounded-md border p-3"
            placeholder="Write a new comment"
            readonly=""
            type="text"
            value=""
          />
        </div>
        <div
          class="flex justify-end gap-2"
        >
          <button
            class="input-disabled cursor-not-allowed bg-gray-200 text-black h-10 rounded-md w-32"
            disabled=""
          >
            Post
          </button>
          <button
            class="h-10 rounded-md border bg-white w-32 text-center text-black hover:cursor-pointer hover:text-white hover:bg-highlight-red transition-all"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;

exports[`Comment Modal: Snapshots > render comment modal not logged in no selected post 1`] = `
<body>
  <div />
</body>
`;

exports[`Comment Modal: Snapshots > render comment modal not logged in with selected post 1`] = `
<body
  aria-hidden="true"
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 max-h-[75vh] w-11/12 max-w-5xl -translate-x-1/2 -translate-y-1/2 overflow-y-auto overflow-x-hidden rounded-md border bg-white p-8 outline-none md:w-1/2"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="mb-4 flex w-full justify-between"
        >
          <div
            class="flex"
          >
            <div
              class="avatar self-center hover:scale-110 transition-all"
            >
               
              <a
                href="/profile/user-1"
              >
                <img
                  alt="avatar"
                  class="inline-block aspect-square rounded-[50%] "
                  src="imageUrl"
                  style="width: 2.5rem;"
                />
              </a>
            </div>
            <div
              class="ml-3"
            >
              <a
                href="/profile/user-1"
              >
                <p
                  class="text-lg font-bold hover:underline"
                >
                  test user
                </p>
              </a>
              <p
                class="text-sm text-gray-400"
              >
                0 seconds ago
              </p>
            </div>
          </div>
        </div>
        <p
          class="mb-3 text-2xl font-bold"
        >
          title
        </p>
        <div
          class="prose max-w-none mb-3"
        >
          <p>
            content
          </p>
        </div>
        <div
          class="flex w-full justify-end border-b mb-3"
        >
          <p
            class="float-right text-highlight-green no-underline hover:no-underline mb-3"
          >
            2
             Comments
          </p>
        </div>
        <div>
          <div
            class="relative m-3 rounded-md border-b p-4"
          >
            <div
              class="flex"
            >
              <div
                class="avatar self-center transition-all hover:scale-110"
              >
                 
                <a
                  href="/profile/user-2"
                >
                  <img
                    alt="avatar"
                    class="inline-block aspect-square rounded-[50%] "
                    src="image"
                    style="width: 2.5rem;"
                  />
                </a>
              </div>
              <div
                class="ml-3"
              >
                <a
                  href="/profile/user-2"
                >
                  <p
                    class="text-lg font-bold hover:underline"
                  >
                    test user 2
                  </p>
                </a>
                <p
                  class="text-gray-400 text-sm"
                >
                  0 seconds ago
                </p>
              </div>
            </div>
            <div
              class="break-words"
            >
              <p
                class="mt-4 w-full"
              >
                test comment
              </p>
            </div>
          </div>
          <div
            class="relative m-3 rounded-md border-b p-4"
          >
            <div
              class="flex"
            >
              <div
                class="avatar self-center transition-all hover:scale-110"
              >
                 
                <a
                  href="/profile/user-3"
                >
                  <img
                    alt="avatar"
                    class="inline-block aspect-square rounded-[50%] "
                    src="image"
                    style="width: 2.5rem;"
                  />
                </a>
              </div>
              <div
                class="ml-3"
              >
                <a
                  href="/profile/user-3"
                >
                  <p
                    class="text-lg font-bold hover:underline"
                  >
                    test user 3
                  </p>
                </a>
                <p
                  class="text-gray-400 text-sm"
                >
                  0 seconds ago
                </p>
              </div>
            </div>
            <div
              class="break-words"
            >
              <p
                class="mt-4 w-full"
              >
                test comment
              </p>
            </div>
          </div>
        </div>
        <div
          class="tooltip tooltip-info w-full"
          data-tip="Sign in to comment"
        >
          <input
            class="input-disabled cursor-not-allowedinput-bordered mb-3 w-full rounded-md border p-3"
            placeholder="Write a new comment"
            readonly=""
            type="text"
            value=""
          />
        </div>
        <div
          class="flex justify-end gap-2"
        >
          <button
            class="input-disabled cursor-not-allowed bg-gray-200 text-black h-10 rounded-md w-32"
            disabled=""
          >
            Post
          </button>
          <button
            class="h-10 rounded-md border bg-white w-32 text-center text-black hover:cursor-pointer hover:text-white hover:bg-highlight-red transition-all"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;
